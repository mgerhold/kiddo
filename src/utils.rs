use std::ffi::OsStr;
use std::path::Path;

use bumpalo::Bump;

use crate::parser::errors::ParserError;
use crate::token::Token;

pub(crate) trait AllocPath {
    fn alloc_path(&self, path: impl AsRef<Path>) -> &Path;
}

impl AllocPath for Bump {
    fn alloc_path(&self, path: impl AsRef<Path>) -> &Path {
        Path::new(unsafe {
            OsStr::from_os_str_bytes_unchecked(
                self.alloc_slice_copy(path.as_ref().as_os_str().as_os_str_bytes()),
            )
        })
    }
}

pub(crate) fn parse_unsigned_int<'a, T: num_traits::Unsigned>(
    token: &'a Token<'a>,
) -> Result<T, ParserError<'a>> {
    let lexeme = token.lexeme();
    if let Some(lexeme) = lexeme.strip_prefix("0x") {
        T::from_str_radix(lexeme, 16).map_err(|_| ParserError::IntegerLiteralOutOfBounds { token })
    } else if let Some(lexeme) = lexeme.strip_prefix("0o") {
        T::from_str_radix(lexeme, 8).map_err(|_| ParserError::IntegerLiteralOutOfBounds { token })
    } else if let Some(lexeme) = lexeme.strip_prefix("0b") {
        T::from_str_radix(lexeme, 2).map_err(|_| ParserError::IntegerLiteralOutOfBounds { token })
    } else {
        T::from_str_radix(lexeme, 10).map_err(|_| ParserError::IntegerLiteralOutOfBounds { token })
    }
}
