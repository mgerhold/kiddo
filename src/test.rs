use utils::setup_test;
use utils::TestFile;

mod utils;

macro_rules! function_name {
    () => {{
        fn f() {}
        fn type_name_of<T>(_: T) -> &'static str {
            std::any::type_name::<T>()
        }
        let name = type_name_of(f);
        name.strip_suffix("::f")
            .unwrap()
            .split("::")
            .last()
            .unwrap()
    }};
}

#[test]
fn empty_source_file_compiles_successfully() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: "",
            },
            &[],
        )
        .compiler_output
    );
}

#[test]
fn importing_symbol_with_the_same_name_as_local_definition_fails() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: r#"from main import f;

export function f() { }
"#,
            },
            &[],
        )
        .compiler_output
    );
}

#[test]
fn importing_symbol_from_other_module() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: r#"from utils import f;
"#,
            },
            &[TestFile {
                filename: "utils.bs",
                source: r#"export function f() { }
"#
            }],
        )
        .compiler_output
    );
}

#[test]
fn global_variables() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: r#"let global0 = 42;
let global1: U32 = 42;
let mutable global2 = 42;
let mutable global3: U32 = 42;
let const global4 = 42;
let const global5: U32 = 42;

export let global6 = 42;
export let global7: U32 = 42;
export let mutable global8 = 42;
export let mutable global9: U32 = 42;
export let const global10 = 42;
export let const global11: U32 = 42;
"#
            },
            &[]
        )
        .compiler_output
    );
}

#[test]
fn local_variables() {
    insta::assert_snapshot!(
        setup_test(
            function_name!(),
            TestFile {
                filename: "main.bs",
                source: r#"function f() {
    let a = 10;
    let const b = 20;
    let mutable c = 30;
    let d: U32 = 40;
    let const e: U32 = 50;
    let mutable f: U32 = 60;
}
"#
            },
            &[]
        )
        .compiler_output
    );
}
