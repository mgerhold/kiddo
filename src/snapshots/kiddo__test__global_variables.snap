---
source: src/test.rs
expression: "setup_test(function_name!(),\n        TestFile {\n            filename: \"main.bs\",\n            source: r#\"let global0 = 42;\nlet global1: U32 = 42;\nlet mutable global2 = 42;\nlet mutable global3: U32 = 42;\nlet const global4 = 42;\nlet const global5: U32 = 42;\n\nexport let global6 = 42;\nexport let global7: U32 = 42;\nexport let mutable global8 = 42;\nexport let mutable global9: U32 = 42;\nexport let const global10 = 42;\nexport let const global11: U32 = 42;\n\"#,\n        }, &[]).compiler_output"
---
[
    ModuleWithConnectedImports {
        filename: "main.bs",
        module: Module {
            imports: [],
            definitions: [
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: false,
                        mutability: Constant,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global0" (main.bs:1:5, byte_offset=4),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: None,
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:1:15, byte_offset=14),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: false,
                        mutability: Constant,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global1" (main.bs:2:5, byte_offset=22),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: Some(
                            Named {
                                name: Relative {
                                    tokens: [
                                        Token {
                                            source_location: "U32" (main.bs:2:14, byte_offset=31),
                                            type_: UppercaseIdentifier,
                                        },
                                    ],
                                },
                            },
                        ),
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:2:20, byte_offset=37),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: false,
                        mutability: Mutable,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global2" (main.bs:3:13, byte_offset=53),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: None,
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:3:23, byte_offset=63),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: false,
                        mutability: Mutable,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global3" (main.bs:4:13, byte_offset=79),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: Some(
                            Named {
                                name: Relative {
                                    tokens: [
                                        Token {
                                            source_location: "U32" (main.bs:4:22, byte_offset=88),
                                            type_: UppercaseIdentifier,
                                        },
                                    ],
                                },
                            },
                        ),
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:4:28, byte_offset=94),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: false,
                        mutability: Constant,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global4" (main.bs:5:11, byte_offset=108),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: None,
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:5:21, byte_offset=118),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: false,
                        mutability: Constant,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global5" (main.bs:6:11, byte_offset=132),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: Some(
                            Named {
                                name: Relative {
                                    tokens: [
                                        Token {
                                            source_location: "U32" (main.bs:6:20, byte_offset=141),
                                            type_: UppercaseIdentifier,
                                        },
                                    ],
                                },
                            },
                        ),
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:6:26, byte_offset=147),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: true,
                        mutability: Constant,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global6" (main.bs:8:12, byte_offset=163),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: None,
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:8:22, byte_offset=173),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: true,
                        mutability: Constant,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global7" (main.bs:9:12, byte_offset=188),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: Some(
                            Named {
                                name: Relative {
                                    tokens: [
                                        Token {
                                            source_location: "U32" (main.bs:9:21, byte_offset=197),
                                            type_: UppercaseIdentifier,
                                        },
                                    ],
                                },
                            },
                        ),
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:9:27, byte_offset=203),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: true,
                        mutability: Mutable,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global8" (main.bs:10:20, byte_offset=226),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: None,
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:10:30, byte_offset=236),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: true,
                        mutability: Mutable,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global9" (main.bs:11:20, byte_offset=259),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: Some(
                            Named {
                                name: Relative {
                                    tokens: [
                                        Token {
                                            source_location: "U32" (main.bs:11:29, byte_offset=268),
                                            type_: UppercaseIdentifier,
                                        },
                                    ],
                                },
                            },
                        ),
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:11:35, byte_offset=274),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: true,
                        mutability: Constant,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global10" (main.bs:12:18, byte_offset=295),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: None,
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:12:29, byte_offset=306),
                                type_: Integer,
                            },
                        ),
                    },
                ),
                GlobalVariable(
                    GlobalVariableDefinition {
                        is_exported: true,
                        mutability: Constant,
                        name: NonTypeIdentifier(
                            Token {
                                source_location: "global11" (main.bs:13:18, byte_offset=327),
                                type_: LowercaseIdentifier,
                            },
                        ),
                        type_: Some(
                            Named {
                                name: Relative {
                                    tokens: [
                                        Token {
                                            source_location: "U32" (main.bs:13:28, byte_offset=337),
                                            type_: UppercaseIdentifier,
                                        },
                                    ],
                                },
                            },
                        ),
                        initial_value: IntegerLiteral(
                            Token {
                                source_location: "42" (main.bs:13:34, byte_offset=343),
                                type_: Integer,
                            },
                        ),
                    },
                ),
            ],
        },
        imports: [],
    },
]
